### PR TITLE
[trainer] feat: support sft_trainer with model engine

### PR DESCRIPTION
**Pull request by vermouth1992**
### What does this PR do?

- support sft_trainer with model engine
- fix engine interface to handle missing data from non-pp
- add gsm8k multi-turn dataset
- add left-right padding to MultiTurnDataset so that the data format of SFT matches with RL
- add sft e2e runnable tests with fsdp and megatron backend

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/trainer/config/sft_trainer_engine.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+# 1. Comments must appear above each field.
+# 2. There must be a blank line between each field.
+# 3. Inline comments (after a field on the same line) are not allowed.
+# 4. Indentation level is respected for nested fields.
+
+# <folder_name>@<field_name>.<field_name>: <yaml_file_name>
+
+defaults:
+  - model@model: hf_model
+  - engine@engine: fsdp
```
Comment (ETOgaosion) [inline verl/trainer/config/sft_trainer_engine.yaml:11]: Would SFT use megatron? Seems that this configuration binds with fsdp.

Comment (vermouth1992) [reply verl/trainer/config/sft_trainer_engine.yaml:11]: this is only the default config. If we set engine=megatron, it will switch to megatron. See the test script

File: verl/utils/dataset/multiturn_sft_dataset.py [hunk 2]
```diff
@@ -301,40 +311,114 @@ def __getitem__(self, item):
...
+                pad_token_id = self.tokenizer.pad_token_id if self.tokenizer.pad_token_id is not None else 0
+                padded_input_ids = torch.full((self.max_length - sequence_length,), pad_token_id, dtype=input_ids.dtype)
+                padded_attention_mask = torch.zeros((self.max_length - sequence_length,), dtype=attention_mask.dtype)
+                padded_loss_mask = torch.zeros((self.max_length - sequence_length,), dtype=loss_mask.dtype)
+
+                input_ids = torch.cat((input_ids, padded_input_ids))
+                attention_mask = torch.cat((attention_mask, padded_attention_mask))
+                loss_mask = torch.cat((loss_mask, padded_loss_mask))
+            elif sequence_length > self.max_length:
+                if self.truncation == "left":
```
Comment (ETOgaosion) [inline verl/utils/dataset/multiturn_sft_dataset.py:339]: What's the difference between `pad_mode` and `truncation`, it there relations with their values?

File: verl/trainer/config/sft_trainer_engine.yaml (hunk 1 above)
Comment (ETOgaosion) [reply verl/trainer/config/sft_trainer_engine.yaml:11]: It's convenient.

File: verl/utils/dataset/multiturn_sft_dataset.py (hunk 2 above)
Comment (vermouth1992) [reply verl/utils/dataset/multiturn_sft_dataset.py:339]: - pad_mode is right or left_right. We use left_right here because currently model engine only supports it. We will implement fully rmpad in trainer so that we won't pad in dataloader as well.
- truncation is we first pad, then truncate. right padding only truncate right while left_right padding truncates left and right.

**Comment by vermouth1992** [comment]
> I think it's OK, maybe some docs are needed to clarify changes, configs and usage?

Yes. We need to write comprehensive doc, docstring and examples to illustrate the core ideas in the following PRs
